Add LZW file support to `io.fits`, `io.ascii` and `utils.data`

kYwzor opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address the lack of LZW support in Astropy. Fixes #10714.

For performance discussion please [see here](https://github.com/astropy/astropy/issues/10714#issuecomment-2766649330). For disclosure, I am the sole developer of [uncompresspy](https://pypi.org/project/uncompresspy/), which I am using in this PR.

There are a few open questions for this PR:

- This implies a new optional dependency for Astropy. I believe I've handled this correctly, but there may be some tricky stuff I missed.
- I consider uncompresspy to still be in beta and thus I'm fully open to any suggestion of changing interfaces, if you think the existing ones are unsatisfying in some way. Ultimately, I built uncompresspy mostly for the Astropy use-case, so I'm willing to adapt whatever needed to best fit it.
- I have not added any tests, though this new feature would surely need some tests. Please advise on how to best handle this.
- I also suspect documentation needs to be updated somewhere to address LZW support.
- The implementation in uncompresspy allows reading of the uncompressed stream byte by byte, without keeping the decompressed buffer in memory. However, in my testing, I've noticed Astropy very frequently seeks the compressed file backwards, even if we're accessing entirely sequential data (e.g. reading header and then data of the first HDU, then reading the header and data of the second HDU, and so forward, will result in Astropy frequently going backwards on the file). This is a huge performance killer since LZW decompression is purely sequential and thus going backwards implies restarting from the very beginning of the file. To reduce this effect, I created a "keep_buffer" option for uncompresspy which makes it so that the LZWFile object keeps an internal buffer of all previously decompressed data, which it leverages to speed up backwards seeks (i.e. avoid repeated decoding). This is frankly unsatisfying, as it results in significantly larger memory consumption for large files, but I'm not sure if there's a better alternative. Very much open to suggestions on this.

mhvk commented:
@kYwzor - before anything else, really nice to have written that new package! Your code changes here look very reasonable, although I'll let @saimn look in more detail. Obviously, we'd want some tests - maybe we can use some of the ESO files you mentioned in the issue? (A smaller one, ideally!).

As for suggestions for how to deal with backwards seeks, since lzw is stream-based, would it be an option to keep some kind of check points so you don't have to start from scratch? Maybe at a given spacing? E.g., I could imagine for FITS files, with their native 2880 byte blocks, to have a dict indexed by uncompressed block giving the corresponding compressed file location and decompressor state. (Obviously, without having looked at your code, *no idea* whether this makes sense!).

Separately, the fact that we apparently skip backward relatively frequently suggests that there are some improvements to make in astropy!

mhvk commented:
p.s. Where does the source code live? I didn't see it on your github page.

kYwzor commented:
@mhvk thanks for the comments. Oops, I forgot to make the repository public, nice catch! It should be public now [here](https://github.com/kYwzor/uncompresspy).

Regarding the test file, yes ESO files are probably fine if they are accessible from the public ESO archive. I did some digging and the file provided by the original poster on #10714 is available on the ESO archive [here](https://archive.eso.org/wdb/wdb/eso/eso_archive_main/query?dp_id=VISIR.2017-01-05T09:04:11.258&format=SexaHour&tab_stat_plot=on&aladin_colour=aladin_instrument) or with a [direct download link here](https://dataportal.eso.org/dataPortal/file/VISIR.2017-01-05T09:04:11.258). The file has about 4MB which I think is reasonable (on my testing uncompresspy needed less than a second to fully decompress it). Let me know if this is what you had in mind.

Agreed on it being puzzling that io.fits backtracks so much; I did not investigate what causes it exactly, I just assumed there was a legitimate reason. Maybe someone can chime in?

As for the checkpoints, that actually sounds like a pretty promising idea. I don't want to "marry" uncompresspy to FITS files specifically, thus tying it to 2880 byte blocks doesn't sound great; however, all LZW files I've found in the wild use the so called "block mode", which means the decompression dictionary gets reset from time to time when a special "CLEAR" code is encountered. A reset of the dictionary means that at that point in the code stream you don't need any knowledge of the prior code stream (other than knowing exactly where the new block starts). Therefore, these would be very "natural" checkpoints to jump to. For the typical file with max code size of 16 bits, CLEAR codes are spaced a **minimum** of 122656 bytes apart (i.e. roughly 120KB, though they typically go a bit beyond that, it varies depending on how good the compression ratio is at that point in the file); this seems good enough to me. This would definitely allow us to go backwards without requiring a full restart of the decompression or forcing us to keep the full decompression buffer in memory. 

I will probably not have the time to attempt an implementation of the checkpoint idea this week, but I could give it a shot in the coming weeks if you think that's worth chasing. I still don't love this approach because we're still decoding previously decoded parts of the file (which is obviously sub-optimal) and thus it could result in significantly worse performance than just keeping the buffer in memory... but I also don't love keeping the buffer in memory, so maybe we really have no alternative and this is the best of two bad options. I guess the only way out is if we could avoid rewinding the file to begin with, but that's likely not possible.

Side note, the `decompress_in_memory` flag on `io.fits` makes it so the entire file is decompressed at once into memory. I wonder if it would be reasonable to have a different flag in which we only decompress what's requested, but we keep in memory whatever has already been decompressed (i.e. lazy decompressing but storing it in memory).

kYwzor commented:
Had another look at the code base. Not familiar with this part of the code, but I suspect we might also need to check for LZW files somewhere around here:

https://github.com/astropy/astropy/blob/2382a9912a025c71eb9802c2efe566457b524a7c/astropy/utils/data.py#L387-L448

I'll try to address this in a later commit.

I also searched for the tests being done for bz2 and it seems they are really spread across a few test files, not entirely sure what the logic here is but I'll leave what I found below:

<details>
<summary>A whole bunch of code snippets</summary>

* Straightforward decompression test for `io.fits`. I could replicate this for LZW but there's an added challenge: we don't really have a way to mimic `_make_bzip2_file` as we can't easily create an LZW file on demand. I guess we would just have to store the file somewhere. Similarly, writing to LZW will just have to be not supported (frankly, why would you want this 😅).
https://github.com/astropy/astropy/blob/2382a9912a025c71eb9802c2efe566457b524a7c/astropy/io/fits/tests/test_core.py#L783-L834

* Straightforward decompression test but for `io.ascii`. My current PR only addresses `io.fits` but it probably could be expanded to `io.ascii` without much effort (though I haven't been able to quickly find exactly where `io.ascii` is handling decompression... might be the get_readable_fileobj I linked above?)
https://github.com/astropy/astropy/blob/2382a9912a025c71eb9802c2efe566457b524a7c/astropy/io/ascii/tests/test_compressed.py#L22-L28

* This one doesn't seem to be testing for bz2 specifically, rather just using it as an example, it doesn't seem like we would need to replicate it to LZW.
https://github.com/astropy/astropy/blob/2382a9912a025c71eb9802c2efe566457b524a7c/astropy/io/ascii/tests/test_read.py#L1353-L1360

* And finally there's a few tests on utils/tests/test_data.py that I don't fully understand the purpose of
https://github.com/astropy/astropy/blob/2382a9912a025c71eb9802c2efe566457b524a7c/astropy/utils/tests/test_data.py#L953-L971
https://github.com/astropy/astropy/blob/2382a9912a025c71eb9802c2efe566457b524a7c/astropy/utils/tests/test_data.py#L996-L1016
https://github.com/astropy/astropy/blob/2382a9912a025c71eb9802c2efe566457b524a7c/astropy/utils/tests/test_data.py#L1172-L1208

</details>

kYwzor commented:
Added support inside get_readable_fileobj, does anyone know an example use-case to test it?

saimn commented:
> Agreed on it being puzzling that io.fits backtracks so much; I did not investigate what causes it exactly, I just assumed there was a legitimate reason. Maybe someone can chime in?

By default we load HDUs lazily and use memmap, so it's not possible to predict if and when data will be loaded.
Maybe try with `memmap=False` (but it depends what you are testing).

> Regarding the test file, yes ESO files are probably fine if they are accessible from the public ESO archive. [...] The file has about 4MB

4MB is too big to store it in the astropy repo, so either we have a smaller file (<50 KB) that we can store here or another option would be to have a remote data test (we probably need to host it on astropy-data instead of using directly the ESO link).

> I wonder if it would be reasonable to have a different flag in which we only decompress what's requested, but we keep in memory whatever has already been decompressed (i.e. lazy decompressing but storing it in memory).

We already a lot of parameters and this one seems a bit too specific. Also it would be probably not apply to other compression methods where we might not control this behavior. 

> I haven't been able to quickly find exactly where io.ascii is handling decompression... might be the get_readable_fileobj I linked above?

Yes I think so.

kYwzor commented:
Regarding the test file, it seems ESO's programmatic access allows you to filter by file size, so I searched for files smaller than 20KB. There's obviously tons, so I just picked the oldest observation (why not!) which is a 13.4KB file (33.7KB uncompressed). This is [the details link](https://archive.eso.org/wdb/wdb/eso/eso_archive_main/query?dp_id=ONTT.1991-12-30T08_55_46.000) and [direct download link](https://dataportal.eso.org/dataPortal/file/ONTT.1991-12-30T08:55:46.000). My thinking is we could put both the compressed and uncompressed files in the repository, so that we could then test the decompression and compare with the uncompressed file. Let me know if this is ok for you.

neutrinoceros commented on pyproject.toml:
```diff
@@ -96,6 +96,7 @@ all = [
     "bottleneck>=1.3.3",
     "fsspec[http]>=2023.4.0",
     "s3fs>=2023.4.0",
+    "uncompresspy"
```
please set a lower bound now so we know where we're starting from
```suggestion
    "uncompresspy>=0.1.1"
```

kYwzor replied:
Thanks, solved this in a later commit

kYwzor commented on astropy/utils/tests/test_data.py:
```diff
@@ -953,11 +953,19 @@ def test_get_invalid(package):
 )
 def test_local_data_obj(filename):
-    if (not HAS_BZ2 and "bz2" in filename) or (not HAS_LZMA and "xz" in filename):
-        with pytest.raises(ValueError, match=r" format files are not supported"):
+    if (
+        (not HAS_BZ2 and "bz2" in filename)
+        or (not HAS_LZMA and "xz" in filename)
+        or (not HAS_UNCOMPRESSPY and "Z" in filename)
+    ):
+        with pytest.raises(
+            ModuleNotFoundError,
+            match=r"(does not provide the [lb]z[2m]a? module\.)|(The optional package uncompresspy is necessary for reading LZW)",
+        ):
```
This change to the "match" condition technically modifies the behavior for the BZ2 and LZMA case. However, I'm fairly sure the previous match was incorrect as I cannot find any "format files are not supported" string. I believe the reason this wasn't raising any errors in testing is because none of our CI tests are missing the LZMA or BZ2 package (maybe we should have that?).

mhvk replied:
I think lzma and bz2 are in the standard library these days - we probably should adjust accordingly, but that's out of scope here.

kYwzor commented:
I've updated uncompresspy and now it is able to keep checkpoints in the file, which enable much better seeking performance. Updated the dependencies and removed the "keep_buffer" option accordingly.

I believe this commit is ready for final review.

kYwzor commented:
Had to make changes to solve merge conflicts resulting from #17968. Commit history was getting really messy so I took the chance to clean it up and squash the commits.

mhvk commented on astropy/io/ascii/tests/test_compressed.py:
```diff
@@ -35,3 +35,12 @@ def test_xz(filename):
+@pytest.mark.parametrize("filename", ["data/short.rdb.Z", "data/ipac.dat.Z"])
+def test_lzw(filename):
+    t_comp = read(get_pkg_data_filename(filename))
+    t_uncomp = read(get_pkg_data_filename(filename.replace(".Z", "")))
```
We only support python >= 3.11, so can use `filename.removesuffix(".Z")`
(I realize all the other tests do the same... so maybe not worth it)

mhvk commented on astropy/io/ascii/tests/test_compressed.py:
```diff
@@ -35,3 +35,12 @@ def test_xz(filename):
+    t_comp = read(get_pkg_data_filename(filename))
+    t_uncomp = read(get_pkg_data_filename(filename.replace(".Z", "")))
+    assert t_comp.dtype.names == t_uncomp.dtype.names
+    assert np.all(t_comp.as_array() == t_uncomp.as_array())
```
Minor since all the other tests do the same, but my tendency would be to do `from numpy.testing import assert_array_equal)` and here do `assert_array_equal(t_comp.as_array(), t_uncomp.as_array())`.

mhvk commented on astropy/io/fits/tests/test_core.py:
```diff
@@ -890,6 +894,69 @@ def test_writeto_lzma_filename(self):
+        is not supported."""
+        filename = self.temp("testname.fits.Z")
+        h = fits.PrimaryHDU()
+        with pytest.raises(OSError):
```
Maybe add `match="mode not supported with LZW files"`

mhvk commented on astropy/utils/data.py:
```diff
@@ -212,7 +213,7 @@ def get_readable_fileobj(
 
     This supports passing filenames, URLs, and readable file-like objects,
-    any of which can be compressed in gzip, bzip2 or lzma (xz) if the
+    any of which can be compressed in gzip, bzip2, lzma (xz) or lzw if the
```
Maybe add `(Z)` after `lzw`? Before this PR, I didn't even know that the compression algorithm for .Z was called LZW...

mhvk reviewed (commented):
Nice! I have only the most nitpicky comments.

kYwzor commented:
@mhvk thanks for the comments, I agree with them and they should be addressed in my latest commit.

Regarding the tests on test_compressed.py, indeed I just copied what the others were doing. I edited my test as you proposed, but kept the others as-is (I think it would be out of scope to touch them).

Regarding lzma and bz2 my understanding is that, indeed they are in the standard library, but minimal Python installations may not include them (or in other words, they are technically an optional part of the Python installation). I think in practice this never happens in Windows, but if you compile from the source in Linux you may be missing some dependencies and thus the support for lzma/bz2 is disabled.

saimn commented on astropy/io/fits/tests/test_core.py:
```diff
@@ -1409,6 +1476,16 @@ def _make_lzma_file(self, filename="test0.fits.xz"):
+            with open(lzw_file, "rb") as f1:
+                with open(temp_file, "wb") as f2:
+                    f2.write(f1.read())
+            return temp_file
```
Since this is just doing a copy you could use `self.copy_file` here.

kYwzor replied:
The challenge is that I need the capability to rename the file (see `test_detect_lzw`). I don't think `self.copy_file` offers that. Maybe we could change the function to allow renames?

saimn replied:
Ah right, yes that's an option, could be done easily, however I'm not sure many tests would need that so the other option would be to do a `shutil.move` in the test that need it. No strong preference on my side so as you prefer :)

kYwzor replied:
I kept it simple and in my latest commit I just changed `_make_lzw_file` to do a copy (and, when a rename is requested, it then moves the file).

Side note, it would probably make a lot of sense for `self.copy_file` to return the path of the copy, since nearly in all cases people do `self.copy_file(filename)` followed by something with `self.temp(filename)`. This would make it easy to add an optional "new_filename" parameter, such that the `_make_lzw_file` could be reduced to simply `return self.copy_file(lzwfile, new_filename)`

saimn replied:
Yes, looks for fine for now, thanks!

kYwzor commented:
Had to do a rebase to fix merge conflicts caused by #17984, should be fine now

mhvk reviewed (approved):
Looks all good to me, but obviously @saimn should approve too.

saimn reviewed (approved):
LGTM, great to see this issue fixed, thanks @kYwzor !
